projects/wallet-lib-client-for-event-queue/0001-event-queue-connection-design.md

```diff
@@ -0,0 +1,199 @@
+[motivation]: #motivation
+
+The event queue is a way to sequentially stream events, making sure we do not miss any transaction or transaction update, we can also continue where we left in case of any connection issues, currently any connection issue makes it so we have to sync the entire wallet history from the beggining.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The fullnode's event queue (a.k.a. reliable integration) has 2 ways to stream events, a websocket and a http api (not server-side events (SSE), just polling events with an event id based pagination).
+We will focus on the websocket implementation but we can later create another class that uses the http api.
+
+## Events
+
+The events from the event queue are not filtered in any way, meaning we have to implement our own filter.
+The events that represent new transactions or blocks is called `NEW_VERTEX_ACCEPTED` and any updates on the transaction data is called `VERTEX_METADATA_CHANGED` these 2 events come with the current transaction data, we can use this data to filter for transaction of our wallet.
+
+## Address subscription
+
+The current implementation uses the fullnode pubsub to listen only for addresses of our wallet, meaning that during startup we send a subscribe command for each address of our wallet.
+Since the events will be filtered locally we can instead use the `subscribeAddresses` method to create a local list of addresses being listened to and use this list to filter the events.
+
+The "list of addresses" will be an object where the address is the key, since determining if an object has a key is O(1) we can ensure that this does not become a bottleneck for wallets with many addresses.
+Alternatively, we could use the storage `isAddressMine` method which is already O(1).
+
+## Event streaming
+
+To get the full state of the wallet we would need to stream all events of the fullnode, but we can still use the address history api to fetch the balance and transactions of our addresses and start listening the newer events.
+
+The best way to achieve this is to use the event api to fetch a single event, this will come with the latest event id.
+Example response of `GET ${FULLNODE_URL}/v1a/event?size=1`
+
+```json
+{
+    "events": [
+        {
+            "peer_id": "ca084565aa4ac6f84c452cb0085c1bc03e64317b64e0761f119b389c34fcfede",
+            "id": 0,
+            "timestamp": 1686186579.306944,
+            "type": "LOAD_STARTED",
+            "data": {},
+            "group_id": null
+        }
+    ],
+    "latest_event_id": 9038
+}
+```
+
+Then we save `latest_event_id` and sync the history with the address history api.
+Once we have the wallet on the current state we can start streaming from the `latest_event_id`.
+There can be transactions arriving during this process which would mean we add them during the history sync and during the event streaming, but this issue does not affect the balance or how we process the history.
```

Comment:
Balance calculation is not affected if the same transaction arrives multiple times.
This sync issue needs to be stated because we have to maintain this characteristic since the alternative would be to sync events from `id = 0` which are millions of events.